README.md
# Password Hasher Plus

[![Build Status](https://travis-ci.org/simu/passwordhasherplus.svg?branch=master)](http://travis-ci.org/simu/passwordhasherplus)

Password Hasher Plus is a replacement for the Password Hasher
extension which unfortunately is not developed anymore and does not
work correctly with multi-process Firefox.

This extension is a port of the Chrome extension of the same name
which generates hash words that are identical to the hash words
generated by Password Hasher, if compatibility mode is enabled.

Unfortunately you will have to migrate site tags and versions manually.

## ChangeLog

* [ChangeLog](CHANGELOG.md)

## Quickstart Guide:

1) First time only: check that site settings are correct and that you have enabled compatibility mode if you're coming from the old Firefox plugin.
1) Input your master password in the password field
1) Press the '#' button near the password field or use Ctrl+# to hash the password in place
1) Optional: Use the '*' button or press Ctrl+* to check either your master password or the hash word (depending on whether you've hashed your master password already or not).
1) Login

## 3rd Party Dependencies

* [tld.js](https://github.com/oncletom/tld.js) v2.3.1
  The file [lib/tld.js](https:/github.com/simu/passwordhasherplus/lib/tld.js)
  is generated by cloning the tld.js repository, checking out the version tag
  (currently v2.3.1) and running `npm run build`. The resulting file is copied
  into the source tree for the extension.
* [jQuery](https://code.jquery.com/jquery-3.1.1.min.js) v3.3.1
  The minified jQuery 3.1.1 javascript is downloaded from code.jquery.com and
  used as-is. The only part of the extension that relies on jQuery is the
  standalone website which is available for users that wish to use
  PasswordHasherPlus outside of Firefox.


CHANGELOG.md
# 2.5.1

Release 2.5.1 includes various small improvements.

## Added

* Option to hide the mask button (#23)

## Changed

* Simplified mask button placement (#25, #28)

## Fixed

* Remove unused CSS class `.hidden` (#26)
* Only start content-script code after options are loaded from storage

# 2.5.0

## Changed

* Revamped workflow: you do not have to enter your password in the website
  anymore.
* Reworked UI: updated design, the popup is now a Page Action instead of a
  Browser Action, and respects the browser UI style.

# 2.4.0

## Changed

* The addon now uses the WebExtensions Storage API with which settings persist.
* You can choose to use your Firefox account to synchronize settings across devices

## Added

* A new password strength option "Custom" allows you to customize hash word
  generation in compatibility mode. This should now make it possible to
  generate hash words which are identical to the old extension with any
  possible combination of flags.
* The portable page now has a button to copy the hash word to the clipboard
* The options page and popup are styled using "browser-style" to look more
  like native UI elements

# 2.3.10

## Fixed

* Fix portable page handling of compatibility mode for individual sites.

# 2.3.9

## Fixed

* Identical to 2.3.8, new patch release due to AMO issues

# 2.3.8

## Fixed

* Fix portable page to properly handle compatibility mode for individual tags
  (i.e. tags with empty seed in configuration, or tags prefixed with
  compatible: if using portable page exclusively)

# 2.3.7

## Fixed

* Code hygiene changes

# 2.3.6

## Added
* Initial release on AMO after forking from ...
* Improved Javascript code for the portable page that doesn't produce errors in the JS console
* Improved GUID generation for private key. Now uses `crypto.getRandomValues()` instead of `Math.random()`
* Fix portable page generation for Firefox
* Fix potential XSS in portable page javascript


tests/test_common.js
var crypto = require('crypto');
var assert = require('assert');
var chai = require('chai');
var expect = chai.expect;

var tld = require('lib/tld.js');

var generateGuid = require('common.js').generateGuid;
var bump = require('common.js').bump;
var grepUrl = require('common.js').grepUrl;
var extractQueryParam = require('common.js').extractQueryParam;
var generateHash = require('common.js').generateHash;

suite('common.js', function() {

    test("generateGuid", function () {
        assert.notEqual(generateGuid(), generateGuid());
    });

    test("bump foo", function () {
        assert.equal(bump("foo"), "foo:1");
    });
    test("bump foo:1", function () {
        assert.equal(bump("foo:1"), "foo:2");
    });
    test("bump foo:100", function () {
        assert.equal(bump("foo:100"), "foo:101");
    });

    test("grepUrl matches simple domain", function () {
        assert.equal(grepUrl("http://foo.com"),"foo");
        assert.equal(grepUrl("https://foo.com"),"foo");
    });

    test("grepUrl matches standard domain", function () {
        assert.equal(grepUrl("http://www.foo.com"),"foo");
    });

    test("grepUrl matches ip addrs", function () {
        assert.equal(grepUrl("http://1.1.1.1"),"1.1.1.1");
        assert.equal(grepUrl("http://111.111.111.111"),"111.111.111.111");
    });

    test("grepUrl matches sub domains", function () {
        assert.equal(grepUrl("http://www2.foo.com"),"foo");
        assert.equal(grepUrl("http://test.foo.com"),"foo");
    });

    test("grepUrl matches uk domains", function () {
        assert.equal(grepUrl("http://www.foo.co.uk"),"foo");
        assert.equal(grepUrl("http://test.foo.co.uk"),"foo");
    });

    test("queryParam finds in 1 param", function () {
        assert.equal(extractQueryParam("http://test.com?foo=true", "foo"),"true");
    });
    test("queryParam finds with no param", function () {
        assert.equal(extractQueryParam("http://test.com", "foo"),null);
        assert.equal(extractQueryParam("http://test.com?a=1", "foo"),null);
    });
    test("queryParam finds wtesth 2 params", function () {
        assert.equal(extractQueryParam("http://test.com?a=1&b=2", "a"),"1");
        assert.equal(extractQueryParam("http://test.com?a=1&b=2", "b"),"2");
    });

});

suite('generateHash', function() {
    var DIGITS = 0;
    var ALPHANUM = 1;
    var SPECIAL = 2;

    var config = {};
    var seed = generateGuid();
    var options = { compatibilityMode: false };
    var input = "mypassword";

    function generate(length, strength) {
        var config = { tag: "mytag", options: options,
            policy: {
                length: length,
                strength: strength,
                seed: seed
            }
        };

        return generateHash(config, input);
    }

    for(i=1; i <= 24; ++i) {
        test("generates "+i+" digit(s)", function () {
            var result = generate(i, DIGITS);
            var result2 = generate(i, DIGITS);
            expect(result.length).to.be.equal(i);
            expect(result).to.match(/^[0-9]+$/);
            expect(result).to.be.equal(result2);
        });

        test("generates "+i+" alphanum(s)", function () {
            var result = generate(i, ALPHANUM);
            var result2 = generate(i, ALPHANUM);
            expect(result.length).to.be.equal(i);
            expect(result).to.match(/^[a-zA-Z0-9]+$/);
            expect(result).to.match(/[0-9]/);
            expect(result).to.be.equal(result2);
        });

        test("generates "+i+" special(s)", function () {
            var result = generate(i, SPECIAL);
            var result2 = generate(i, SPECIAL);
            expect(result.length).to.be.equal(i);
            expect(result).to.match(/^[a-zA-Z0-9!@#$%^&*()\\\/+'",.-]+$/);
            expect(result).to.match(/[!@#$%^&*()\\\/+'",.-]/);
            expect(result).to.be.equal(result2);
        });
    }

    test("ignores seed in compatibility mode", function () {
        var withSeed = generate(24, SPECIAL);
        var result2 = generate(24, SPECIAL);
        options.compatibilityMode = true;
        var withoutSeed = generate(24, SPECIAL);
        seed = generateGuid();
        var withNewSeed = generate(24, SPECIAL);

        expect(withSeed).to.not.be.equal(withoutSeed);
        expect(withoutSeed).to.be.equal(withNewSeed);
        expect(withSeed).to.be.equal(result2);
    });

    test("compatible with old extension - custom strength, twitter+123456", function() {
        var cstm = { "d": false, "p": false, "m": false, "r": false };
        var twitter_cfg = {
            tag: "twitter", options: { compatibilityMode: true },
            policy: { length: 8, strength: -1, custom: cstm }
        };
        var twitter_old = 'scxqJ/Lx';
        var twitter_gen = generateHash(twitter_cfg, "123456");
        assert.equal(twitter_gen, twitter_old);
    });
    test("compatible with old extension - custom strength, twitter+aaaaaa", function() {
        var cstm = { "d": false, "p": false, "m": false, "r": false };
        var twitter_cfg = {
            tag: "twitter", options: { compatibilityMode: true },
            policy: { length: 8, strength: -1, custom: cstm }
        };
        var twitter_old = 'q6oCYnFI';
        var twitter_gen = generateHash(twitter_cfg, "aaaaaa");
        assert.equal(twitter_gen, twitter_old);
    });
});


popup.js
var storage;
var url;
var config;
var debug = true;

function rehash(e) {
    var masterpw = document.getElementById('masterpw').value;
    var hash = generateHash (config, masterpw);
    // same hint algo as passhash-ng
    var master_hint = PassHashCommon.generateHashWord(
            masterpw, masterpw, 2, true, false, true, true, false);
    // generate hashword hint based on config, but length 2
    if (debug) console.log("config.policy.length: "+config.policy.length);
    var hash_hint = generateHashL(config, masterpw, 2);
    document.getElementById('hashword').value = hash;
    document.getElementById('master-hint').textContent = master_hint;
    document.getElementById('hash-hint').textContent = hash_hint;
}

function writeModel () {
    config.tag = document.getElementById('tag').value;
    if (config.tag.startsWith ("compatible:")) {
        config.tag = config.tag.substringAfter ("compatible:");
        delete config.policy.seed;
    } else {
        config.policy.seed = config.options.privateSeed;
    }
    config.policy.length = document.getElementById('length').value;
    config.policy.strength = document.getElementById('strength').value;
    if (config.policy.strength == -1) {
        config.policy.custom = new Object();
        config.policy.custom.d = document.getElementById('d').checked;
        config.policy.custom.p = document.getElementById('p').checked;
        config.policy.custom.m = document.getElementById('m').checked;
        config.policy.custom.r = document.getElementById('r').checked;
    } else {
        delete config.policy.custom;
    }
    if(null == config.policy.seed || config.policy.seed == config.options.privateSeed) {
        document.getElementById('syncneeded').classList.add("hidden");
    }
    if (debug) console.log("[popup.js] saving config");
    storage.saveConfig(url, config, refreshPopup);
}

function readModel () {
    if (debug) console.log("readModel()");
    document.getElementById('tag').value = config.tag;
    document.getElementById('length').value = config.policy.length;
    document.getElementById('strength').val = config.policy.strength;
    if (true == config.options.compatibilityMode) {
        document.getElementById('compatmodeheader').innerHTML =
            "<b>Compatibility:</b> on";
    } else if (null == config.policy.seed) {
            document.getElementById('tag').value = "compatible:" + config.tag;
    }
    if (false == config.options.backedUp && false == config.options.compatibilityMode) {
        document.getElementById('compatmodeheader').innerHTML =
            "<b>Warning:</b> You have not yet indicated that you have backed up your private key. Please do so on the Options page.";
    }
    if(null != config.policy.seed && config.policy.seed != config.options.privateSeed) {
        document.getElementById('syncneeded').classList.remove('hidden');
    }
    if (config.policy.strength == -1) {
            if (debug) console.log("custom strength: show checkboxes");
            document.getElementById('strength-requirements').classList.remove('hidden');
            if (config.policy.custom === undefined) {
                config.policy.custom = config.options.custom;
            }
            document.getElementById('d').checked = config.policy.custom.d;
            document.getElementById('p').checked = config.policy.custom.p;
            document.getElementById('m').checked = config.policy.custom.m;
            document.getElementById('r').checked = config.policy.custom.r;
    } else {
            if (debug) console.log("not custom strength: hide checkboxes");
            document.getElementById('strength-requirements').classList.add('hidden');
    }
    if (debug) console.log("[popup.js:readModel] rehashing...");
    rehash();
}

function refreshPopup() {
    chrome.tabs.query({active: true, currentWindow: true}, function (tabs) {
        url = chrome.extension.getBackgroundPage ().grepUrl (tabs[0].url);
        if (debug) console.log('loading/creating config for url='+url);
        storage.loadConfig(url, (cfg) => {
            if (debug) console.log('got config='+JSON.stringify(cfg, null, 2));
            config = cfg;
            config.fields = toSet (config.fields);
            readModel ();
        });
    });
}

document.getElementById('bump').addEventListener('click', function () {
    var tagField = document.getElementById('tag');
    tagField.value = bump (tagField.value);
    writeModel ();
});

document.getElementById('tag').addEventListener('input', writeModel);
document.getElementById('tag').addEventListener('change', writeModel);
document.getElementById('length').addEventListener('change', writeModel);
document.getElementById('strength').addEventListener('change', writeModel)
document.getElementById('d').addEventListener('change', writeModel);
document.getElementById('p').addEventListener('change', writeModel);
document.getElementById('m').addEventListener('change', writeModel);
document.getElementById('r').addEventListener('change', writeModel);
// for hashing
document.getElementById('masterpw').addEventListener('input', rehash);
document.getElementById('masterpw').addEventListener('change', rehash);
// form submit / cancel
document.getElementById('cancel').addEventListener('click', function() {
    window.close();
});
document.getElementById('hasher').addEventListener('submit', function() {
    var hash = document.getElementById('hashword').value;
    if(debug) console.log("[popup.js] submitting hash '" + hash + "' to content script");
    browser.extension.getBackgroundPage().forwardHash(config.tag, hash);
    window.close();
});

/* handle masking / unmasking of password and hash fields */
function maskUnmask(elemid) {
    var buttonid = 'unmask-'+elemid;
    var masked = document.getElementById(elemid).type == 'password';
    if (masked) {
        document.getElementById(elemid).type = 'text';
        document.getElementById(buttonid).textContent = '*';
    } else {
        document.getElementById(elemid).type = 'password';
        document.getElementById(buttonid).textContent = 'a';
    }
}

document.getElementById('unmask-masterpw').addEventListener('click',
        ()=>{maskUnmask('masterpw')});
document.getElementById('unmask-hashword').addEventListener('click',
        ()=>{maskUnmask('hashword')});

// populate popup fields
refreshPopup();

document.addEventListener('DOMContentLoaded', function() {
    document.getElementById('link-options').addEventListener('click', function() {
        chrome.runtime.openOptionsPage();
    });
    document.getElementById('portablePage').addEventListener('click', function() {
        // For compatibility with Firefox just do /page.html for URL
        chrome.tabs.create({url:'/passhashplus.html?tag=' + document.getElementById('tag').value})
    });
});


common.js
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is Password Hasher Plus
 *
 * The Initial Developer of the Original Code is Eric Woodruff.
 * Portions created by the Initial Developer are Copyright (C) 2010
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s): (none)
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

var debug = false;

// XXX: Inject npm packages when running tests
if (typeof require !== 'undefined') {
    if (typeof tld === 'undefined') {
        var tld = require('lib/tld.js');
    }
    if (typeof crypto === 'undefined') {
        var crypto = new Object();
        crypto.getRandomValues = require('get-random-values');
    }
    if (typeof PassHashCommon === 'undefined') {
        PassHashCommon = require('lib/passhashcommon.js').PassHashCommon;
    }
} else {
    // tldjs 2.3.1 exports itself as window.tldjs
    window.tld = window.tldjs
}

String.prototype.startsWith = function (str) {
	return (this.match ("^" + str) == str);
}

String.prototype.substringAfter = function (str) {
	return (this.substring (this.indexOf (str) + str.length));
}

var Set = function () {}
Set.prototype.add = function (o) { this[o] = true; }
Set.prototype.remove = function (o) { delete this[o]; }

function toSet (array) {
	var s = new Set ();
	for (var i = 0; i < array.length; ++i) {
		s.add (array[i]);
	}
	return s;
}

function toArray (s) {
	return Object.keys (s);
}

var default_length = 8;
var default_strength = 2;
var default_hashkey = "Ctrl+Shift+51";
var default_maskkey = "Ctrl+Shift+56";
var default_sync = false;

function generateGuid () {
	var template = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
	var xycount = (template.match(/[xy]/g) || []).length;
	var rand = new Uint8Array(xycount);
	crypto.getRandomValues(rand);
	var i = 0;
	return template.replace (/[xy]/g, function(c) {
		var r = rand[i++] % 16, v = c == 'x' ? r : (r&0x3|0x8);
		return v.toString (16);
	}).toUpperCase ();
}

function generateHashL (config, input, length) {
    if (debug) console.log("[common.js:generateHashL] length = " + length);
    var tag = config.tag;

    if (false == config.options.compatibilityMode && null != config.policy.seed) {
        tag = PassHashCommon.generateHashWord (
                config.policy.seed,
                tag,
                24,
                true, // require digits
                true, // require punctuation
                true, // require mixed case
                false, // no special characters
                false // only digits
                );
    }

    if (config.policy.strength == -1) {
        return PassHashCommon.generateHashWord (
                tag,
                input,
                length,
                config.policy.custom.d, // require digits
                config.policy.custom.p, // require punctuation
                config.policy.custom.m, // require mixed case
                config.policy.custom.r, // no special characters
                false // only digits
                );
    } else {
        return PassHashCommon.generateHashWord (
                tag,
                input,
                length,
                true, // require digits
                config.policy.strength > 1, // require punctuation
                true, // require mixed case
                config.policy.strength < 2, // no special characters
                config.policy.strength == 0 // only digits
                );
    }
}

function generateHash (config, input) {
    return generateHashL(config, input, config.policy.length);
}

function bump (tag) {
	var re = new RegExp ("^([^:]+?)(:([0-9]+))?$");
	var compatible = false;
	if (tag.startsWith ("compatible:")) {
		tag = tag.substringAfter ("compatible:");
		compatible = true;
	}
	var matcher = re.exec (tag);
	var bump = 1;
	if (null != matcher[3]) {
		tag = matcher[1];
		bump += parseInt (matcher[3]);
	}
	if (compatible) {
		tag = "compatible:" + tag;
	}
	return tag + ":" + bump;
}

function select_storage_area() {
    return browser.storage.local.get('sync').then(results => {
        var sync = 'sync' in results && results['sync'];
        if (sync) {
            return browser.storage.sync;
        } else {
            return browser.storage.local;
        }
    });
}

function dumpDatabase() {
    return select_storage_area().then(area => {
        return area.get(null).then(results=>{
            var sane_order_results = new Object();
            sane_order_results.version = results.version;
	    sane_order_results.sync = results.sync;
            sane_order_results.options = results.options;
            sane_order_results.url = results.url;
            sane_order_results.tag = results.tag;
            return sane_order_results;
	});
    });
}

/* grepUrl:
   take a url and return the corresponding site tag.

   (a) split the url into parts. Discard everything except the address.
       e.g. http://www.google.com:80/mail -> www.google.com

   (b) If the address looks like an ip address or dot free hostname, then
       return it. It can be used as the site tag.
       e.g. 192.168.1.1 -> 192.168.1.1
            localhost -> localhost

   (c) Any address which does not meet (b)'s criteria is probably a domain.
       Use the tldjs library (http://github.com/oncletom/tldjs) to:
         * Cut off the registrar controled part at the end of the domain
         * Cut off any subdomains off the beginning of the domain

       e.g. google.com -> google
            google.co.uk -> google
            mail.google.co.uk -> google
            calendar.google.co.uk -> google

       This is an important step because, (using google as an example):
       The passwords used at google.com, google.co.uk, google mail,
       and google calendar, all need to be the same.

   (d) If the URL is formatted in any other way, fall back to the default
       "chrome" site tag
*/

function grepUrl (url) {
    var split_at_first_dot = /(^[^.]+)\..*$/;
    var split_url = /^(https?:\/\/)(.+@)?([^:#\/]+)(:\d{2,5})?(\/.*)?$/;
    // 1 = protocol, 2 = auth, 3 = address, 4 = port, 5 = path
    // split_url is stolen from http://github.com/oncletom/tldjs
    var is_ipv4 = /^\d{1,3}(\.\d{1,3}){3}$/;
    var is_dot_free_hostname = /^[^.]+$/;

    try {
        //if url badly formed, this will throw a type error, handled at (d)
        var address = split_url.exec(url)[3];                          // a
        if (is_ipv4.test(address) || is_dot_free_hostname.test(address)) {
            return address;                                            // b
        } else {
            //this shouldn't throw an error.
            //but just in case it does, handle it at (d)
            console.log("grepUrl: address=" + address);
            console.log("grepUrl: getDomain=" + tld.getDomain(address));
            return split_at_first_dot.exec(tld.getDomain(address))[1]; // c
        }
    } catch (e) {
        console.log(e)
        return "chrome";                                               // d
    }
}

function extractQueryParam(string, key){
    key = key.replace(/[*+?^$.\[\]{}()|\\\/]/g, "\\$&"); // escape RegEx meta chars
    var match = string.match(new RegExp("[?&]"+key+"=([^&]+)(&|$)"));
    return match && decodeURIComponent(match[1].replace(/\+/g, " "));
}

// http://stackoverflow.com/questions/7731778/jquery-get-query-string-parameters
function queryParam(key) {
    return extractQueryParam(location.search, key);
}

if (typeof exports !== 'undefined') {
    exports.bump = bump;
    exports.generateGuid = generateGuid;
    exports.grepUrl = grepUrl;
    exports.extractQueryParam = extractQueryParam;
    exports.generateHash = generateHash;
}

function copy() {
  var copyText = document.querySelector("#hash");
  copyText.select();
  document.execCommand("Copy");
}



background.js
/*
 * We fire change and keyup events after we edit the value of the currently
 * focused password field to catch the most common ways to do client-side
 * password verification.
 */
function updateFocusedField(tabid, hash) {
    browser.tabs.executeScript(tabid, {
        code: `
document.activeElement.value = '${hash}';
var evt = new Event('change');
setTimeout(document.activeElement.dispatchEvent(evt), 0);
evt = new Event('keyup');
setTimeout(document.activeElement.dispatchEvent(evt), 0);`
    });
}

function forwardHash(tag, hash) {
    if (debug) console.log("[background.js:forwardHash] got hash " + hash + " for page " + tag);
    browser.tabs.query({active:true,currentWindow:true}).then(results => {
        var tab = results[0];
        updateFocusedField(tab.id, hash);
    });
}

/* enable page action for all regular tabs */
browser.tabs.query({}).then(opentabs => {
    for (let tab of opentabs) {
        browser.pageAction.show(tab.id);
    }
});

/* enable page action for new tabs */
browser.tabs.onUpdated.addListener((tabid,changeinfo) => {
    browser.pageAction.show(tabid);
});

/* create context menu item for all editable fields */
browser.menus.create({
    id: 'open-passhash',
    title: 'Password Hasher Plus',
    /* TODO: figure out icons */
    icons: {
        "16": "images/passhash.png",
        "32": "images/passhash.png"
    },
    contexts: ["editable", "password"],
    command: "_execute_page_action",
});


options.js
// TODO:
// notify user if they have local and sync settings that differ + offer to
// merge
var storage;

function setNewGuid () {
	var seedElement = document.getElementById ("seed");
	seedElement.value = generateGuid ();
}

function refreshOptionsPage(tag) {
    if (debug) console.log("[options.js] refreshOptionsPage called: " + tag);
    // make sure checkbox does not remain checked
    document.getElementById("sync-options").value = 0;
    restoreOptions();
    refreshStorage();
}

function saveSync() {
    var sync = document.getElementById("sync").value == 1;
    var syncopt = document.getElementById("sync-options").value;
    if (debug) console.log("[options.js:saveSync] sync=" + sync);
    if (debug) console.log("[options.js:saveSync] syncopt=" + syncopt);
    storage.migrateArea(sync, syncopt, () => { refreshOptionsPage('saveSync') });
}

// saveOptions does not update sync status, see saveSync().
function saveOptions () {
	var options = new Object ();
	options.defaultLength = document.getElementById ("length").value;
	options.defaultStrength = document.getElementById ("strength").value;
	console.log("defaultStrength = " + options.defaultStrength);
	if (options.defaultStrength == -1) {
		options.custom = new Object();
		options.custom.d = document.getElementById('d').checked;
		options.custom.p = document.getElementById('p').checked;
		options.custom.m = document.getElementById('m').checked;
		options.custom.r = document.getElementById('r').checked;
	} else {
		delete options.custom;
	}
	options.showMaskButton = document.getElementById ("maskbutton").checked;
	options.compatibilityMode = document.getElementById ("compatibility").checked;
	options.privateSeed = document.getElementById ("seed").value;
	options.backedUp = document.getElementById ("backedup").checked;
	options.hashKey = document.getElementById ("hashkey").value;
	options.maskKey = document.getElementById ("maskkey").value;
        if (debug) console.log('[options.js] Saving options='+JSON.stringify(options,null,2));
        storage.saveOptions(options, () => { refreshOptionsPage('saveOptions') });
}

function restoreOptions () {
    storage.loadOptions((options) => {
        document.getElementById ("length").value = options.defaultLength;
        document.getElementById ("strength").value = options.defaultStrength;
        document.getElementById ("compatibility").checked = options.compatibilityMode;
        document.getElementById ("maskbutton").checked = options.showMaskButton;
        document.getElementById ("seed").value = options.privateSeed;
        document.getElementById ("backedup").checked = options.backedUp;
        document.getElementById ("hashkey").value = options.hashKey;
        document.getElementById ("maskkey").value = options.maskKey;
	if (debug) console.log('setting sync value to '+ options.sync);
        document.getElementById ("sync").value = options.sync ? 1 : 0;
    });
}



function refreshStorage() {
	dumpDatabase().then(db => {
		document.getElementById ("everything").value = JSON.stringify(db, null, 2);
	});
}

function clearStorage () {
	if (confirm ("You are about to erase all of the Password Hasher Plus database. " +
		    "This is typically done before loading a snapshot of a previous database state. " +
		    "Are you certain you want to erase the database?")) {
		select_storage_area().then(area => {
                    area.clear ();
                    alert ("Your database is now empty. " +
                            "You probably want to paste a previous snapshot of the database to the text area to the right, " +
                            "and hit \"Load\" to re-populate the database. " +
                            "Good luck.");
                });
	}
	storage.migrate ();
}

function loadStorage () {
	try {
		everything = JSON.parse(document.getElementById ("everything").value);
	} catch(e) {
		alert("Sorry, the data in the text area to the right is not valid JSON.");
		return;
	}
        // clear and rewrite storage area
        storage.init(everything).then(() => { refreshOptionsPage('load from db') });
}

function setShortcut(action, e) {
	if (e.which == 16 || e.which == 17)
		return;
	if (action == "hash")
		hk = document.getElementById('hashkey');
	if (action == "mask")
		hk = document.getElementById('maskkey');
	if (e.which != 0)
		hk.value = (e.ctrlKey ? "Ctrl+" : "") + (e.shiftKey ? "Shift+" : "") + e.which;
	else
		hk.value = (action == "hash" ? "Ctrl+Shift+51" : "Ctrl+Shift+56");
}

function showHideCustomStrength() {
  var strength = document.getElementById('strength').value;
  if (strength == -1) {
    document.getElementById('customstrengthrow').classList.remove("hidden");
  } else {
    document.getElementById('customstrengthrow').classList.add("hidden");
  }
}
// Add event listeners once the DOM has fully loaded by listening for the
// `DOMContentLoaded` event on the document, and adding your listeners to
// specific elements when it triggers.
document.addEventListener('DOMContentLoaded', function () {
    restoreOptions ();
    refreshStorage ();

    // make sure that when storage is changed, and options page is open, we
    // reflect the changed settings.
    browser.storage.onChanged.addListener(function(changes, area) {
        restoreOptions();
        refreshStorage();
    });

    document.getElementById('generate').addEventListener('click', setNewGuid);
    document.getElementById('backupSave').addEventListener('click', saveOptions);
    document.getElementById('backupRevert').addEventListener('click', restoreOptions);
    document.getElementById('removeUnUsedTags').addEventListener('click',
            function() {storage.collectGarbage (); refreshStorage ();});
    document.getElementById('dbClear').addEventListener('click',
            function() {clearStorage (); refreshStorage ();});
    document.getElementById('dbSave').addEventListener('click', loadStorage);
    document.getElementById('dbRevert').addEventListener('click', refreshStorage);
    document.getElementById('syncSave').addEventListener('click', saveSync);
    document.getElementById('strength').addEventListener('change', showHideCustomStrength);

    document.getElementById('portablePage').addEventListener('click',
            function() { chrome.tabs.create({url:'/passhashplus.html'}) });
	
    document.getElementById('hashkey').addEventListener('keydown',
            function(e) {setShortcut("hash", e)});
    document.getElementById('maskkey').addEventListener('keydown',
            function(e) {setShortcut("mask", e)});
    document.getElementById('haskeydefault').addEventListener('click',
            function() {var e = new Object(); e.which=0; setShortcut("hash", e)});
    document.getElementById('maskkeydefault').addEventListener('click',
            function(e) {var e = new Object(); e.which=0; setShortcut("mask", e)});

});
